### PR TITLE
launch fighters

### PR DESCRIPTION
**Pull request by NickAragua**
This creates a specialized function that lets the bot launch fighters from carriers at a "proper" rate (rather than one per turn, sticking to the 'safe' launch rate of two fighters per bay door per turn).

Also changes the bot's "can this unit move this turn" criteria to match exactly what Server thinks they should be, to fix a game stopper bug that occurs when the bot is using battle armor/infantry with the "infantry don't count for movement initiative" setting.

### CODE REVIEW

File: megamek/src/megamek/client/bot/princess/Princess.java [hunk 1]
```diff
@@ -2098,6 +2105,55 @@ private void unloadTransportedInfantry(MovePath path) {
...
     }
     
+    /**
+     * Helper function that adds an "launch" step for units that are transporting 
+     * launchable units in some kind of bay.
+     */
+    private void launchFighters(MovePath path) {
+        // if my objective is to cross the board, even though it's tempting, I won't be leaving the infantry
+        // behind. They're not that good at screening against high speed pursuit anyway.
+        if(getBehaviorSettings().shouldGoHome()) {
```
Comment (Windchild292) [inline megamek/src/megamek/client/bot/princess/Princess.java:2115]: Spacing... Please try to fix this part of your style, as it is very inaccessible

File: megamek/src/megamek/client/bot/princess/Princess.java [hunk 2]
```diff
@@ -2098,6 +2105,55 @@ private void unloadTransportedInfantry(MovePath path) {
...
+            return;
+        }
+        
+        Entity movingEntity = path.getEntity();
+        Coords pathEndpoint = path.getFinalCoords();
+        Targetable closestEnemy = getPathRanker(movingEntity).findClosestEnemy(movingEntity, pathEndpoint, getGame(), false);
+
+        // if there are no enemies on the board, then we're not launching anything.
+        if((null == closestEnemy) || (closestEnemy.getTargetType() != Targetable.TYPE_ENTITY)) {
```
Comment (Windchild292) [inline megamek/src/megamek/client/bot/princess/Princess.java:2124]: spacing

File: megamek/src/megamek/client/bot/princess/Princess.java [hunk 3]
```diff
@@ -2098,6 +2105,55 @@ private void unloadTransportedInfantry(MovePath path) {
...
+            return;
+        }
+        
+        TreeMap<Integer, Vector<Integer>> unitsToLaunch = new TreeMap<>();
```
Comment (Windchild292) [inline megamek/src/megamek/client/bot/princess/Princess.java:2128]: Can just be a map

File: megamek/src/megamek/client/bot/princess/Princess.java [hunk 4]
```diff
@@ -2098,6 +2105,55 @@ private void unloadTransportedInfantry(MovePath path) {
...
+        boolean executeLaunch = false;
+        
+        // loop through all fighter (or smallcraft) bays in the current entity
+        // grouping launched craft by bay to limit launches to 'safe' rate.
+        Vector<Bay> fighterBays = movingEntity.getFighterBays();
+        
+        for(int bayIndex = 0; bayIndex < fighterBays.size(); bayIndex++) {
```
Comment (Windchild292) [inline megamek/src/megamek/client/bot/princess/Princess.java:2135]: spacing

File: megamek/src/megamek/client/bot/princess/Princess.java [hunk 5]
```diff
@@ -2098,6 +2105,55 @@ private void unloadTransportedInfantry(MovePath path) {
...
+            Bay bay = fighterBays.get(bayIndex);
+            
+            for(Entity loadedEntity : bay.getLaunchableUnits()) {
```
Comment (Windchild292) [inline megamek/src/megamek/client/bot/princess/Princess.java:2138]: spacing

File: megamek/src/megamek/client/bot/princess/Princess.java [hunk 6]
```diff
@@ -2098,6 +2105,55 @@ private void unloadTransportedInfantry(MovePath path) {
...
+                unitsToLaunch.putIfAbsent(bayIndex, new Vector<Integer>());
```
Comment (Windchild292) [inline megamek/src/megamek/client/bot/princess/Princess.java:2139]: There's almost certainly no need to specify it as integer

**Review by Windchild292 — CHANGES_REQUESTED** [review]
Inaccessible spacing on multiple if and for loops

File: megamek/src/megamek/client/bot/princess/Princess.java (hunk 3 above)
Comment (NickAragua) [reply megamek/src/megamek/client/bot/princess/Princess.java:2129]: Nope, it's passed to a method that expects a TreeMap

**Comment by NickAragua** [comment]
Undid the removal of the "isSimultaneous" check. Turns out I also had simultaneous movement on, which Princess doesn't support anyway.

**Review by Windchild292 — APPROVED** [review]

**Pull request merged**